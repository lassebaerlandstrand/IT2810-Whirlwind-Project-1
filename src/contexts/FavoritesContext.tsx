import React, { createContext, ReactNode, useContext, useEffect, useState } from 'react';
import { Location } from '../types/api-types';

interface FavoritesContextProps {
  favorites: Location[];
  toggleFavorite: (location: Location) => void;
}

const FavoritesContext = createContext<FavoritesContextProps | undefined>(undefined);

export const useFavorites = () => {
  const context = useContext(FavoritesContext);
  if (!context) {
    throw new Error('useFavorites must be used within a FavoritesProvider');
  }
  return context;
};

export const FavoritesProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [favorites, setFavorites] = useState<Location[]>(() => {
    const storedFavorites = localStorage.getItem('favorites');
    return storedFavorites ? JSON.parse(storedFavorites) : [];
  });

  useEffect(() => {
    localStorage.setItem('favorites', JSON.stringify(favorites));
  }, [favorites]);

  const toggleFavorite = (location: Location) => {
    setFavorites((prevFavorites) => {
      const isFavorite = prevFavorites.some(
        (fav) => fav.city_name === location.city_name && fav.country_name === location.country_name,
      );
      if (isFavorite) {
        return prevFavorites.filter(
          (fav) => fav.city_name !== location.city_name || fav.country_name !== location.country_name,
        );
      } else {
        return [...prevFavorites, location];
      }
    });
  };

  return <FavoritesContext.Provider value={{ favorites, toggleFavorite }}>{children}</FavoritesContext.Provider>;
};
