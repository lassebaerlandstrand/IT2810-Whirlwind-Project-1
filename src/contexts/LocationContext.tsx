import React, { createContext, ReactNode, useContext, useState } from 'react';
import { Location } from '../types/api-types';
import LOCATIONS from '../utils/locations';

type OptionsType = {
  [key: string]: (a: Location, b: Location) => number;
};

export const options: OptionsType = {
  'City Name': (a: Location, b: Location) => a.city_name.localeCompare(b.city_name),
  'Country Name': (a: Location, b: Location) => a.country_name.localeCompare(b.country_name),
};

interface LocationContextProps {
  sortedLocations: Location[];
  setSortCondition: (sortCondition: () => (a: Location, b: Location) => number) => void;
}

const LocationContext = createContext<LocationContextProps | undefined>(undefined);

export const LocationProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [sortedLocations, setSortedLocations] = useState<Location[]>(LOCATIONS);
  const [sortCondition, setSortCondition1] = useState<() => (a: Location, b: Location) => number>(
    () => Object.values(options)[1],
  );

  // const sortLocations = useCallback(() => {
  //   let sortedData = [...LOCATIONS];
  //   if (sortCondition) {
  //     sortedData.sort(sortCondition());
  //   }
  //   setSortedLocations(sortedData);
  // }, [sortCondition]);

  // useEffect(() => {
  //   sortLocations();
  // }, [sortCondition, sortLocations]);

  const setSortCondition = () => (sortCondition: (a: Location, b: Location) => number) => {};

  return <LocationContext.Provider value={{ sortedLocations, setSortCondition }}>{children}</LocationContext.Provider>;
};

export const useLocations = () => {
  const context = useContext(LocationContext);
  if (!context) {
    throw new Error('useLocations must be used within a LocationProvider');
  }
  return context;
};
