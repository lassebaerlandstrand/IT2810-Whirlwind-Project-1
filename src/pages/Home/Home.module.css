.homeContainer {
  display: flex;
  padding: 1.5rem;
  flex-direction: column;
  align-items: center;
  max-width: var(--breakpoint-md);
  margin: auto;
  gap: 1rem;
}

.searchSortContainer {
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  gap: 1rem;
  width: 100%;
}
