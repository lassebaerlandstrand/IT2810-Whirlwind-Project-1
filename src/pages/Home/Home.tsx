import Button from '../../components/Button/Button';
import Example from '../../components/Example/Example';

const Home = () => {
  return (
    <div>
      <h1>43 Â°C</h1>
      <h1>ğŸ”¥ğŸ˜°ğŸ”¥</h1>
      <Button text="Press me to do absolutely nothing" />
      <Example />
    </div>
  );
};

export default Home;
