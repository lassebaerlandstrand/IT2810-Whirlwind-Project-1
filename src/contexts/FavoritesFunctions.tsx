import { useContext } from 'react';
import { FavoritesContext } from './FavoritesContext';

// Had to extract to a serparate file to fix eslint warning
export const useFavorites = () => {
  const context = useContext(FavoritesContext);
  if (!context) {
    throw new Error('useFavorites must be used within a FavoritesProvider');
  }
  return context;
};
