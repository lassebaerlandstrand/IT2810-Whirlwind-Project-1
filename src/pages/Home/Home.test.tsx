import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { LocationProvider } from '../../contexts/LocationContext';
import Home from './Home';

const queryClient = new QueryClient();

describe('Home', () => {
  test('should match snapshot', () => {
    const { asFragment } = render(
      <QueryClientProvider client={queryClient}>
        <MemoryRouter>
          <LocationProvider>
            <Routes>
              <Route path="/" element={<Home />} />
            </Routes>
          </LocationProvider>
        </MemoryRouter>
      </QueryClientProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
