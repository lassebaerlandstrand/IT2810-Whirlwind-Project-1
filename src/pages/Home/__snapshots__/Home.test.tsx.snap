// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > should match snapshot 1`] = `
<DocumentFragment>
  <main
    class="_homeContainer_da09e4"
  >
    <header>
      <a
        class="_header_63c2ea"
        href="/"
      >
        <img
          alt="Whirlwind Logo"
          class="_img_63c2ea"
          src="/project1/src/assets/Whirlwind_Logo.png"
        />
        <div>
          <h1>
            Whirlwind
          </h1>
          <h2>
            Weather Forecast
          </h2>
        </div>
      </a>
    </header>
    <div
      class="_searchSortContainer_da09e4"
    >
      <label
        class="_label_ea9aeb"
      >
        Search cities:
        <div
          class="_searchBarContainer_ea9aeb"
          role="search"
        >
          <span
            class="_searchIcon_ea9aeb"
            data-testid="search-icon"
          >
            <svg
              class="tabler-icon tabler-icon-search "
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 10m-7 0a7 7 0 1 0 14 0a7 7 0 1 0 -14 0"
              />
              <path
                d="M21 21l-6 -6"
              />
            </svg>
          </span>
          <input
            class="_searchInput_ea9aeb"
            placeholder="City Name..."
            type="text"
            value=""
          />
        </div>
      </label>
      <menu
        class="_container_20b9ed"
      >
        <label
          class="_label_20b9ed"
          for=":r0:"
        >
          Sort by:
        </label>
        <button
          aria-controls=":r1:"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="_dropDownButton_20b9ed "
          data-testid="dropdown-button"
          id=":r0:"
        >
          City Name 
          <span
            class="_iconSpan_20b9ed"
          >
            <svg
              aria-hidden="true"
              class="tabler-icon tabler-icon-chevron-down "
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 9l6 6l6 -6"
              />
            </svg>
          </span>
        </button>
        <ul
          aria-labelledby=":r0:"
          class="_dropDownList_20b9ed "
          data-open="false"
          id=":r1:"
          role="listbox"
        >
          <li
            aria-selected="false"
            class="_dropDownItem_20b9ed"
            role="option"
          >
            Country Name
          </li>
        </ul>
      </menu>
    </div>
    <div
      class="_weatherList_4fc390"
      data-testid="weather-list"
    >
      <section
        class="_section_4fc390"
      >
        <h2>
          Other Locations
        </h2>
        <ul
          class="_list_4fc390"
        >
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
          <li>
            <div
              class="_loading_2ba686"
            >
              <svg
                class="tabler-icon tabler-icon-rotate-clockwise _rotating_2ba686"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.05 11a8 8 0 1 1 .5 4m-.5 5v-5h5"
                />
              </svg>
               Loading...
            </div>
          </li>
        </ul>
      </section>
    </div>
  </main>
</DocumentFragment>
`;
